lang/java/avro/src/main/java/org/apache/avro/generic/PrimitivesArrays.java

```diff
@@ -0,0 +1,609 @@
+    }
+
+    @Override
+    public void add(int location, Float o) {
+      if (o == null) {
+        return;
+      }
+      this.add(location, o.floatValue());
+    }
+
+    public void add(int location, float o) {
+      if (location > size || location < 0) {
+        throw new IndexOutOfBoundsException("Index " + location + " out of bounds.");
+      }
+      if (size == elements.length) {
+        // Increase size by 1.5x + 1
+        final int newSize = size + (size >> 1) + 1;
+        elements = Arrays.copyOf(elements, newSize);
+      }
+      System.arraycopy(elements, location, elements, location + 1, size - location);
+      elements[location] = o;
+      size++;
+    }
+
+    @Override
+    public Float set(int i, Float o) {
+      if (o == null) {
+        return null;
+      }
+      return this.set(i, o.floatValue());
+    }
+
+    public float set(int i, float o) {
+      if (i >= size)
+        throw new IndexOutOfBoundsException("Index " + i + " out of bounds.");
+      float response = elements[i];
+      elements[i] = o;
+      return response;
+    }
+
+    @Override
+    public Float remove(int i) {
+      if (i >= size)
+        throw new IndexOutOfBoundsException("Index " + i + " out of bounds.");
+      float result = elements[i];
+      --size;
+      System.arraycopy(elements, i + 1, elements, i, (size - i));
+      return result;
+    }
+
+    @Override
+    public Float peek() {
+      return (size < elements.length) ? elements[size] : null;
+    }
+
+    @Override
+    protected void swap(final int index1, final int index2) {
+      float tmp = this.get(index1);
+      this.set(index1, this.get(index2));
+      this.set(index2, tmp);
+    }
+  }
+
+  public static class DoubleArray extends GenericData.AbstractArray<Double> {
+    private static final double[] EMPTY = new double[0];
+
+    private double[] elements = EMPTY;
+
+    public DoubleArray(int capacity, Schema schema) {
+      super(schema);
+      if (!Schema.Type.DOUBLE.equals(schema.getElementType().getType()))
+        throw new AvroRuntimeException("Not a double array schema: " + schema);
+      if (capacity != 0)
+        elements = new double[capacity];
+    }
+
+    public DoubleArray(Schema schema, Collection<Double> c) {
+      super(schema);
+      if (c != null) {
+        elements = new double[c.size()];
+        addAll(c);
+      }
+    }
+
+    @Override
+    public void clear() {
+      size = 0;
+    }
+
+    @Override
+    public Double get(int i) {
+      return this.getDouble(i);
+    }
+
+    /**
+     * Direct primitive int access.
+     * 
+     * @param i : index.
+     * @return value at index.
+     */
+    public double getDouble(int i) {
+      if (i >= size)
+        throw new IndexOutOfBoundsException("Index " + i + " out of bounds.");
+      return elements[i];
+    }
+
+    @Override
+    public void add(int location, Double o) {
+      if (o == null) {
+        return;
+      }
+      this.add(location, o.floatValue());
```

Comment:
This introduces a loss of precision, since it converts the `Double` to a `float`, and then casts back to `double`. An example:
```java
jshell> Double d = 40.001
d ==> 40.001
jshell> double d2 = d.floatValue()
d2 ==> 40.000999450683594
```
So for example, if you have a field in the schema that's an array of doubles, and you use the Avro JSON parser to read a record with that schema from a JSON object where the value of that field is `[40.001]`, the value in the resulting object will be an array with the element `40.000999450683594`. This caused failures in unit tests that were using `Precision.equals` to compare values because that difference is much bigger than `Precision`'s default epsilon.

Reply:
It looks like Confluent Schema Registry may also be running into this regression